Change register_autograd to reflect ordering of setup_context and backward

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124496
* #124495
* #124414
* __->__ #124403
* #124199
* #124134
* #124299
* #124200
* #124180

old: `register_autograd(setup_context, backward, /)`
new: `register_autograd(backward, /, *, setup_context=None)`

Motivations:
- We introduce these APIs as "give us a backward and use setup_context
  to save things for backward".
- setup_context isn't always necessary.

Test Plan:
- tests

zou3519 committed:
Change register_autograd to reflect ordering of setup_context and backward

old: `register_autograd(setup_context, backward, /)`
new: `register_autograd(backward, /, *, setup_context=None)`

Motivations:
- We introduce these APIs as "give us a backward and use setup_context
  to save things for backward".
- setup_context isn't always necessary.

Test Plan:
- tests

[ghstack-poisoned]

albanD reviewed (approved):
SGTM
Skipping bc-linter since this was added just below in the stack and we don't need to document this bc-breaking change.

zou3519 committed:
Update on "Change register_autograd to reflect ordering of setup_context and backward"


old: `register_autograd(setup_context, backward, /)`
new: `register_autograd(backward, /, *, setup_context=None)`

Motivations:
- We introduce these APIs as "give us a backward and use setup_context
  to save things for backward".
- setup_context isn't always necessary.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change register_autograd to reflect ordering of setup_context and backward"


old: `register_autograd(setup_context, backward, /)`
new: `register_autograd(backward, /, *, setup_context=None)`

Motivations:
- We introduce these APIs as "give us a backward and use setup_context
  to save things for backward".
- setup_context isn't always necessary.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 committed:
Update on "Change register_autograd to reflect ordering of setup_context and backward"


old: `register_autograd(setup_context, backward, /)`
new: `register_autograd(backward, /, *, setup_context=None)`

Motivations:
- We introduce these APIs as "give us a backward and use setup_context
  to save things for backward".
- setup_context isn't always necessary.

Test Plan:
- tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
